src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs

```diff
@@ -0,0 +1,180 @@
+    internal class DesignTimeInputsFileWatcher : ProjectValueDataSourceBase<string>, IVsFreeThreadedFileChangeEvents2, IDesignTimeInputsFileWatcher
+    {
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IDesignTimeInputsDataSource _designTimeInputsDataSource;
+        private readonly IVsService<IVsAsyncFileChangeEx> _fileChangeService;
+        private ImmutableDictionary<string, uint> _fileWatcherCookies = ImmutableDictionary<string, uint>.Empty.WithComparers(StringComparers.Paths);
+
+        private int _version;
+        private readonly DisposableBag _disposables = new DisposableBag();
+
+        /// <summary>
+        /// The block that receives updates from the active tree provider.
+        /// </summary>
+        private IBroadcastBlock<IProjectVersionedValue<string>>? _broadcastBlock;
+
+        /// <summary>
+        /// The public facade for the broadcast block.
+        /// </summary>
+        private IReceivableSourceBlock<IProjectVersionedValue<string>>? _publicBlock;
+
+        [ImportingConstructor]
+        public DesignTimeInputsFileWatcher(ConfiguredProject project,
+                                           IDesignTimeInputsDataSource designTimeInputsDataSource,
+                                           IVsService<SVsFileChangeEx, IVsAsyncFileChangeEx> fileChangeService)
+             : base(project.Services, synchronousDisposal: true, registerDataSource: false)
+        {
+            _threadingService = project.Services.ThreadingPolicy;
+            _designTimeInputsDataSource = designTimeInputsDataSource;
+            _fileChangeService = fileChangeService;
+        }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(DesignTimeInputsFileWatcher));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<string>> SourceBlock
+        {
+            get
+            {
+                EnsureInitialized();
+
+                return _publicBlock!;
+            }
+        }
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            IDisposable dataSourceLink = _designTimeInputsDataSource.SourceBlock.LinkToAsyncAction(ProcessDesignTimeInputs);
+
+            IDisposable join = JoinUpstreamDataSources(_designTimeInputsDataSource);
+
+            _disposables.AddDisposable(dataSourceLink);
+            _disposables.AddDisposable(join);
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<string>>(nameFormat: nameof(DesignTimeInputsFileWatcher) + "Broadcast {1}");
+            _publicBlock = _broadcastBlock.SafePublicize();
+        }
+
+        private async Task ProcessDesignTimeInputs(IProjectVersionedValue<DesignTimeInputs> input)
+        {
+            // We don't want to watch any new files while we're disposing
+            if (IsDisposing)
```

Comment:
But if you're already disposed its okay? What if you get disposed while this method is running?

Reply:
See my comment below in dispose to handle this, both methods should be protected by a lock.